pkg/processor/worker/worker.go

```diff
@@ -160,25 +188,22 @@ func (w *Worker) Terminate() error {
 }
 
 func (w *Worker) Drain() error {
-	w.drainedLock.Lock()
-	defer w.drainedLock.Unlock()
-
-	if !w.isDrained.Load() {
-		err := w.runtime.Drain()
-		if err == nil {
-			w.logger.DebugWith("Successfully drained worker", "workerIndex", w.index)
-			w.isDrained.Store(true)
-		}
+	w.state.Lock()
+	defer w.state.Unlock()
+
+	if err := w.runtime.Drain(); err == nil {
```

Comment:
If the state is unlocked but is set to `Drained`, we will drain again, because you removed the `isDrained` flag from the runtime as well.
I think we should bring back the check to not drain twice in this case